sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java

```diff
@@ -179,13 +179,17 @@ static Context mergeRequestOptionsContext(Context context, RequestOptions option
     }
 
     static Flux<ByteBuffer> validateLength(final HttpRequest request) {
-        final Flux<ByteBuffer> bbFlux = request.getBody();
-        if (bbFlux == null) {
+        if (request.getBody() == null) {
             return Flux.empty();
         }
 
-        final long expectedLength = Long.parseLong(request.getHeaders().getValue("Content-Length"));
+        final Flux<ByteBuffer> bbFlux = request.getBody()
+            .flatMap(buffer -> {
+                // This is done to ensure that the buffer can be reread if possible.)
+                return Flux.just(buffer.duplicate());
+            });
```

Comment:
1. is there some way of being more selective with applying this logic? if there's no chance flux will be re-read this is going to increase allocation overhead. I.e. do this in policies that require it?
2. Would "resetting" buffer work instead of allocation new wrappers?
3. Consider 'map' instead of 'flatMap' - might be less allocation.
4. Should we move this logic to HttpRequest or HttpClient ? validateLength is non-obvious place for this to live in.

Reply:
Thank you for the feedback. I would love to get more eyes on this one.

1. It is hard to tell when we would get into this state because it depends on what causes a retry. Looking at the documentation allocation overhead was limited to book-keeping capacity, position, mark etc. while the content is shared, is that not the case?
2. We could consider reset however I was given the feedback in a separate implementation that we prefer duplicate over manually resetting as it is cleaner so followed the same principle here.
3. Sure.
4.  I went back and forth on this one and decided on this for the following - validatelength is the method that is causing the read so it should be the one responsible for resetting it as well.

Reply:
re 1. and 2. There's very similar logic in the Storage's retry policy so adding this here is going to make storage calls double-allocate. With similar code in the Storage we recall cases where if source buffers were small the allocation overhead could consume up to 2x of what actually buffers carry. I think that to make it right we should simultaneously remove similar logic from Storage (and check if there are other places already doing it) as well as figure out if we can also add "BufferedFlux" in a way how BufferedInputStream works (i.e. passtrhough for buffers large enough and squash them otherwise).
Btw. I think that implementing "reset" would require to keep long pointers in the closure so it's as expensive as duplicating . You're right about this.

re 4. I wouldn't say that validate length triggers read here. it just adds transformations, the read is triggered by subscription to the flux somewhere else. Assuring that buffers are re-readable and validating length aren't concepts that are very related. When thinking where this belongs - why wouldn't we want to benefit from it if somebody calls httppipeline directly (here we're assuming that it's convinient place because of how our stack works). And lastly, the length validation is likely going to change when sync stack gets introduced so placing this in the HttpRequest (or maybe a policy that's added implicitly) and decoupling this from rest proxy is going to help.

Reply:
Yeah, this logic to duplicate should be in HttpRequest. I am thinking each call to `httpRequest.getBody()` should return a duplicate and if we really want the original buffer, we can add a method to do that (don't add this method eagerly). This will also help DPG requests.

re 1, I think this logic should belong in core and all client libraries (like storage) already duplicating the buffer should be updated to let core handle it.

@pallavit Assuming storage and other client libraries are not handling the duplication, are there any other scenarios where we want to let the client library control whether the buffer should be duplicated or not? If there are scenarios like that, we can add a context key to selectively duplicate the buffer.

Reply:
> Assuming storage and other client libraries are not handling the duplication, are there any other scenarios where we want to let the client library control whether the buffer should be duplicated or not? If there are scenarios like that, we can add a context key to selectively duplicate the buffer.

I am not aware of any other scenarios right now.

Reply:
I'd add this in HttpRequest ctor. We call for body in quite many places along the way from api to network - each get would allocate new operators.

Reply:
Since, we want to work with storage team on this and we are close to the release I am going to remove this change from core and add it to the ACR redirect policy to unblock both ACR and core releases. I will create a follow up item to make these changes in core for next release train. Let me know if anyone disagrees.

Also,  please review the OkHttpClient changes as I would like those to go in now. They seem far more harmless.

Reply:
Done.